Dockerfile

```diff
@@ -1,32 +1,21 @@
 FROM node:4.4
-
-ENV user nodegoat_docker
-ENV workdir /usr/src/app/
-
+ENV WORKDIR /usr/src/app/
+WORKDIR $WORKDIR
+COPY package*.json $WORKDIR
+RUN npm install --production --no-cache
+
+FROM node:4-alpine
+ENV USER nodegoat
+ENV GROUP nodegoat
+ENV WORKDIR /home/$USER/app
 # Home is required for npm install. System account with no ability to login to shell
-RUN useradd --create-home --system --shell /bin/false $user
-
-RUN mkdir --parents $workdir
-WORKDIR $workdir
-COPY package.json $workdir
-
+RUN addgroup -S $GROUP && adduser -S -D -s /bin/false $USER -G $GROUP
+WORKDIR $WORKDIR
+COPY --from=0 /usr/src/app/node_modules node_modules
+ADD https://raw.githubusercontent.com/eficode/wait-for/master/wait-for /usr/local/bin/wait-for
```

Comment:
@binarymist how about we use something like https://www.npmjs.com/package/wait-on ?
we can add it to the package manifest for nodegoat, or even better spawn it as a one off with `npx wait-on 'something'`

I understand we're taking a chance to trust on third-parties but there has to be some level of trust anyway, worst-case if one of those deps get deprecated we can just replace it with something else.

Reply:
I like `npx`, but it removes an opportunity for the likes of snyk to analyse the package manifest (I think), and although it's kind of a slick way of running arbitary code, by the look of it, I don't think it's any better than adding and invoking https://raw.githubusercontent.com/eficode/wait-for/master/wait-for . Thoughts?

Reply:
What about using the wait-for function itself as a standalone code inside `entrypoint.sh` and remove the wait-for usage ? its just a for loop anyway as you can see in here:
```
  for i in `seq $TIMEOUT` ; do
    nc -z "$HOST" "$PORT" > /dev/null 2>&1
    
    result=$?
    if [ $result -eq 0 ] ; then
      if [ $# -gt 0 ] ; then
        exec "$@"
      fi
      exit 0
    fi
    sleep 1
  done
  echo "Operation timed out" >&2
  exit 1
```

Reply:
RE npx and `wait-on` - I don't think there's any expectation to scan those. They are helpers in building the docker image and having an environment setup, not part of the application itself.

I think it is more likely for the gist to go offline than the npm package to get removed.
If we still want to keep the bash script then I'd copy it (with whatever license and copyright notice it has) to our own repo and use as a helper through the build, without needing to hit github's gists for it.

Reply:
So what is the last decision regarding this matter? Should we go with offline wait-for script ?

Reply:
I'm not sure if @ckarande has snyk running on NodeGoat or not. I don't see any difference in running arbitrary code in an environment setup as opposed to running arbitrary code as part of the application. Either way, it's still arbitrary code execution on someone's system.

If it was my decision, I'd have something reputable checking for [A9](https://www.owasp.org/index.php/Top_10-2017_A9-Using_Components_with_Known_Vulnerabilities)'s, whether it was environmental or app code. I don't think this is my decision to make though. I'd like to see @ckarande make this decision known.

I think a npm package makes better sense than a script that's not in npm in terms of availability, but personally, I'd be happier still if the few lines were pulled into NodeGoat, but then if we have something like snyk checking the package manifest, then it's slightly less of an issue.

A decision from @ckarande is probably required here.

Reply:
@lirantal @binarymist @ckarande I have removed wait-for and used a normal bash command instead, i think this resolves this conversation as we don't need to worry about using 3rd party script or npm module, I have tested it and works as expected